main.py

```diff
@@ -0,0 +1,169 @@
+
+
+def get_conn(host, port, service, user, password):
+    """ Create oracle database connection """
+    # make connect descriptor string
+    oracle_data_source_name = cx_Oracle.makedsn(host, port, service_name=service)
+    # create and return oracle connection object
+    return cx_Oracle.connect(user=user, password=password, dsn=oracle_data_source_name)
+
+
+def get_oracle_db_records():
+    """
+    connects to oracle db and executes query, converts rows retrieved into dicts
+    :return: list of records (dict)
+    """
+    conn = get_conn(HOST, PORT, SERVICE, USER, PASSWORD)
+    cursor = conn.cursor()
+    cursor.execute(QUERY)
+
+    # define row handler which returns each row as a dict
+    # h/t https://stackoverflow.com/questions/35045879/cx-oracle-how-can-i-receive-each-row-as-a-dictionary
+    cursor.rowfactory = lambda *args: dict(
+        zip([d[0] for d in cursor.description], args)
+    )
+    rows = cursor.fetchall()
+    conn.close()
+
+    logging.info(f"{len(rows)} records in oracle database received.")
+    return rows
+
+
+def generate_record_id(call_number, timestamp):
+    """
+    the records from the oracle database do not have record ids associated with them
+    the coldfusion app added IDs generated by Coldfusion using #hash(call_number,"SHA")#
+    this function does the same using hashlib and appending our timestamp as before
+    :param call_number: call_number
+    :param timestamp:
+    :return: traffic_incident_id string
+    """
+    # create coldfusion hash id
+    cf_id = hashlib.sha1(str.encode(call_number)).hexdigest().upper()
+    #  compose record id from entry identifier (which is not wholly unique) and publication timestamp
+    return "{}_{}".format(cf_id, timestamp)
+
+
+def get_active_records():
+    """
+    Query active records from postgrest endpoint
+    :return: list of active records (dict)
+    """
+    active_records_endpoint = f"{PGREST_ENDPOINT}?traffic_report_status=eq.ACTIVE"
+
+    active_records_response = requests.get(active_records_endpoint, headers=headers)
+    active_records_response.raise_for_status()
+    return active_records_response.json()
+
+
+def format_record(incident):
+    """
+    Formats data from incident into record dict to upsert to postgrest
+    :param incident: dict from oracle db
+    :return: record dict
+    """
+    record = {}
+    published_date = arrow.get(incident['CURR_DATE']).replace(tzinfo="US/Central")
+    status_date = arrow.now().format()
+    record["traffic_report_id"] = generate_record_id(incident['CALL_NUMBER'], published_date.timestamp)
+    record["published_date"] = published_date.format()
+    record["traffic_report_status"] = "ACTIVE"
+    record["traffic_report_status_date_time"] = status_date
+    record["address"] = incident['ADDRESS'].strip()
+    record["issue_reported"] = incident['DESCRIPTION'].strip()
+    record["latitude"] = incident['LATITUDE']
+    record["longitude"] = incident['LONGITUDE']
```

Comment:
I'm also just curious about what additional fields they may have that we could augment the dataset with

Reply:
there is also "BEAT" "AGENCY TYPE" "PRIORITY" and sometimes "CROSS STREET"

Reply:
ooooh i kinda want to scoop those up...even if we don't publish them 😈

Reply:
I'd be for scooping them up too. I'm realizing I don't know how to add new fields to postgrest without using the knack-services metadata_to_postgrest script